src/decorator.ts

```diff
@@ -0,0 +1,55 @@
+import { isObservable, Observable } from 'rxjs';
+
+import { untilDestroyed } from './take-until-destroy';
+
+/**
+ * Automatically unsubscribes from pipe when component destroyed.
+ *
+ * @example
+ * ```ts
+ * @Component({...})
+ * class MyComponent implements OnDestroy {
+ *   @WithUntilDestroyed()
+ *   stream$ = new Observable(...);
+ *
+ *   // OnDestroy method is required by Angular Compiler
+ *   ngOnDestroy() {}
+ * }
+ * ```
+ *
+ * Uses {@link untilDestroyed} operator on the pipe.
+ *
+ * Do not forget to implement {@link OnDestroy} life-cycle hook.
+ */
+export function WithUntilDestroyed(
+  destroyMethodName?: string,
+): PropertyDecorator {
+  return (target, propKey) => {
+    let val: Observable<any>;
+
+    function getter() {
+      return val;
+    }
+
+    function setter(newVal) {
+      if (isObservable(newVal)) {
+        val = newVal.pipe(untilDestroyed(target, destroyMethodName));
+      } else {
+        throw Error(
+          `WithUntilDestroyed: Property ${String(propKey)} on ${
+            target.constructor.name
+          } is not Observable!`,
+        );
+      }
+    }
+
+    if (delete target[propKey]) {
```

Comment:
You don't need to delete when you're returning new descriptor.

Reply:
I do need to delete because I'm redefining new property with same name.